### PR TITLE
Improve Pipeline logs error states

### PR DESCRIPTION
**Pull request by manaswinidas**
JIRA: 
[RHOAIENG-254](https://issues.redhat.com/browse/RHOAIENG-254)
[RHOAIENG-255](https://issues.redhat.com/browse/RHOAIENG-255)

500 Internal Error state will be handled by [RHOAIENG-1067](https://issues.redhat.com/browse/RHOAIENG-1067)

## Description

1. Adds error messages for failed and cleaned-up pods for Pipeline logs
2. Adds error message in case there is a network issue
3. Hides the Logs toolbar toolbar for such states

Cleaned-up pods(Error message according to this [Slack message](https://redhat-internal.slack.com/archives/C05SMJ09DD2/p1701364992500169?thread_ts=1701266219.319199&cid=C05SMJ09DD2)):

Screenshot 2023-12-19 at 7 07 08 PM

No internet(Error message according to this [Slack message](https://redhat-internal.slack.com/archives/C05SMJ09DD2/p1701364343573709?thread_ts=1701242954.097309&cid=C05SMJ09DD2)):

Screenshot 2024-01-15 at 5 19 15 PM

Failed pod(Error message according to this [Slack message](https://redhat-internal.slack.com/archives/C05SMJ09DD2/p1704998689792069?thread_ts=1701266219.319199&cid=C05SMJ09DD2)):

Screenshot 2024-01-15 at 8 14 15 PM

## How Has This Been Tested?

1. Click on any running pipeline with a failed node and check the error state in the Logs Tab
2. Click on any node of an old pipeline(pods are cleaned-up) and check the error state
4. Check the error state for no-internet state too.

## Request review criteria:

Self checklist (all need to be checked):
- [X] The developer has manually tested the changes and verified that the changes work
- [X] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [X] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has added tests or explained why testing cannot be added (unit tests & storybook for related changes)

If you have UI changes: 

- [X] Included any necessary screenshots or gifs if it was a UI change.
- [X] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

### CODE REVIEW

**Comment by manaswinidas** [comment]
WIP to add tests and screenshots for cleaned-up pods.

**Comment by manaswinidas** [comment]
@yih-wang can you check the error states?

**Comment by yih-wang** [comment]
@manaswinidas No, I don't think we will include any error state in the toolbar. The only two ways we display the error messages are:
1. From the inline alert as what you show in the screenshot - that will show the pod error or some other general issue, e.g. network issue
2. From the step dropdown - that will specify which step is encountering the error, offering users detailed insights into the exact point of failure for more effective troubleshooting

**Comment by yih-wang** [comment]
Oops sorry that I misread the question... No we don't show toolbar for both cases

**Comment by manaswinidas** [comment]
@yih-wang it's highly unlikely that some step logs may be fetched while some cannot, in case of network issues.

**Comment by yih-wang** [comment]
@manaswinidas Then I think we should still show the toolbar in the 'No internet' error case to provide the ability to switch to other steps.

**Comment by manaswinidas** [comment]
@yih-wang But we can't retrieve logs when there is no internet connection, even if we are able to switch the steps using the dropdown or the download dropdown, it's doing nothing because there is no internet. Here's a screen recording to demonstrate the same. Do we still show the toolbar in this case?

https://github.com/opendatahub-io/odh-dashboard/assets/22885912/41abaa1a-e7c2-4049-979a-543f89d53b8c

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (Gkrumbach07) [inline frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx:36]: I would abstract that `error?.message.includes('Failed to fetch') ...` logic into some variable or function above in this component because it is being used multiple times in this component. The inline logic is confusing to read otherwise

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (Gkrumbach07) [inline frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx:41]: refresh calls refresh here but really doesnt do anything it seems? You are polling which does the refresh, but the button doesnt actually make a network call

**Review by Gkrumbach07 — COMMENTED** [review]
on error, you are still polling, i assume this is intended?

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (manaswinidas) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx:41]: I agree. @yih-wang do we need a "Retry" action link here? This doesn't do anything.
Screenshot 2024-01-11 at 3 19 42 PM

Comment (yih-wang) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: @manaswinidas I didn't get the reason why we can't have a button and allow users to try to connect internet again through that.

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: That is better, but i still get lost in the rest of the logic. This is fine for now so i can resolve this comment, but for better readability i would suggest something like this. This component is just this logs and error so you can afford to have a little logic in the main body of the component
```tsx
    const isNetworkError = error?.message.includes('Failed to fetch');
    const podCondition = podStatus === null || (podError && podStatus?.completed);

    let alertTitle;
    let alertMessage;

    if (isNetworkError) {
        alertTitle = `${error?.message} logs`;
        alertMessage = 'Check your connection and reload the page.';
    } else if (podCondition) {
        alertTitle = podError?.message;
        alertMessage = `${podName} may have been pruned to prevent over-utilization of resources.`;
    }
```

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (Gkrumbach07) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: I think it can be in there. But because the logs would poll fairly quickly, it doesn't really serve a purpose. I also noticed that for some reason, even when refresh was called, a network call would not be made.

Comment (yih-wang) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: Am I correct in understanding that even in the presence of a connection issue, the log viewer will continue attempting to fetch logs from the server every 2 seconds? This essentially serves the same purpose as a manual refresh button, right?

 If that's the case, I agree that we could remove the retry button and change the alert description to something like "Check your network connection. The system will recover the connection when the network is restored."

Comment (manaswinidas) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: @yih-wang 

> Am I correct in understanding that even in the presence of a connection issue, the log viewer will continue attempting to fetch logs from the server every 2 seconds? This essentially serves the same purpose as a manual refresh button, right?

Yes, log viewer attempts to fetch logs every 3 seconds. Yes, this acts as a manual refresh button.

> If that's the case, I agree that we could remove the retry button and change the alert description to something like "Check your network connection. The system will recover the connection when the network is restored."

Thanks, this works.

Comment (manaswinidas) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: Updated, thanks.

File: frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx
Comment (manaswinidas) [reply frontend/src/concepts/pipelines/content/pipelinesDetails/pipelineRun/runLogs/LogsTabStatus.tsx]: Oh yes, thanks for pointing this out.

**Comment by manaswinidas** [comment]
@yih-wang Can you have a final look at the screenshots?

**Comment by yih-wang** [comment]
@manaswinidas Didn't we combine case 1 (failed/cleaned-up pods) and case 3 (failed pod) you show in the screenshots?

**Comment by manaswinidas** [comment]
@yih-wang I did [according to this discussion ](https://redhat.enterprise.slack.com/archives/C05SMJ09DD2/p1703060015139629?thread_ts=1701266219.319199&cid=C05SMJ09DD2) we had last month

**Comment by manaswinidas** [comment]
@yih-wang Thanks for pointing it out. I updated it just now.

**Comment by Gkrumbach07** [comment]
code looks good to me.

/lgtm

just need @yih-wang approval and an advisor

**Comment by manaswinidas** [comment]
We have @yih-wang approval [here](https://github.com/opendatahub-io/odh-dashboard/pull/2300#issuecomment-1893249712). She was just asking me to change the PR description as it was outdated.

**Review by mturley — APPROVED** [review]

**Comment by manaswinidas** [comment]
Rebased, cleaned up a few nits after the last merge.

**Pull request merged**
